src/mongocxx/collection.cpp

```diff
@@ -189,66 +189,111 @@ stdx::optional<result::bulk_write> collection::bulk_write(const class bulk_write
-        } else if (*options.cursor_type() == cursor::type::k_tailable) {
-            query_flags |= ::MONGOC_QUERY_TAILABLE_CURSOR;
+        if (*options.cursor_type() == cursor::type::k_tailable) {
+            options_builder << "tailable" << bsoncxx::types::b_bool{true};
         } else if (*options.cursor_type() == cursor::type::k_tailable_await) {
-            query_flags |= ::MONGOC_QUERY_TAILABLE_CURSOR;
-            query_flags |= ::MONGOC_QUERY_AWAIT_DATA;
+            options_builder << "tailable" << bsoncxx::types::b_bool{true};
+            options_builder << "awaitData" << bsoncxx::types::b_bool{true};
+        } else if (*options.cursor_type() == cursor::type::k_non_tailable) {
         } else {
             throw logic_error{error_code::k_invalid_parameter};
         }
     }
 
-    filter_bson.init_from_static(filter_builder.view());
+    if (options.hint()) {
+        options_builder << "hint" << options.hint()->to_value();
+    }
 
-    const mongoc_read_prefs_t* rp_ptr = NULL;
+    if (options.limit()) {
+        options_builder << "limit" << *options.limit();
+    }
 
-    if (options.read_preference()) {
-        rp_ptr = options.read_preference()->_impl->read_preference_t;
+    if (options.max()) {
+        options_builder << "max" << *options.max();
+    }
+
+    if (options.max_scan()) {
+        options_builder << "maxScan" << *options.max_scan();
+    }
+
+    if (options.max_time()) {
+        options_builder << "maxTimeMS" << bsoncxx::types::b_int64{options.max_time()->count()};
+    }
+
+    if (options.min()) {
+        options_builder << "min" << *options.min();
+    }
+
+    if (options.no_cursor_timeout()) {
+        options_builder << "noCursorTimeout" << *options.no_cursor_timeout();
+    }
+
+    if (options.projection()) {
+        options_builder << "projection" << bsoncxx::types::b_document{*options.projection()};
+    }
+
+    if (options.return_key()) {
+        options_builder << "returnKey" << *options.return_key();
+    }
+
+    if (options.show_record_id()) {
+        options_builder << "showRecordId" << *options.show_record_id();
+    }
+
+    if (options.snapshot()) {
+        options_builder << "snapshot" << *options.snapshot();
+    }
+
+    if (options.skip()) {
+        options_builder << "skip" << *options.skip();
+    }
+
+    if (options.sort()) {
+        options_builder << "sort" << bsoncxx::types::b_document{*options.sort()};
+    }
+
+    return options_builder.extract();
+}
+
+}  // namespace
+
+cursor collection::find(view_or_value filter, const options::find& options) {
+    // libmongoc::collection_find_with_opts does not support the legacy "modifiers" options, so we
+    // must copy the options struct and convert all of the modifiers options to their modern
+    // equivalents.
+    options::find options_converted = options.convert_all_modifiers();
+
+    scoped_bson_t filter_bson{std::move(filter)};
+    scoped_bson_t options_bson{build_find_options_document(options_converted)};
```

Comment:
Mmm, I think that's not correct.  `options_bson` is generated from `build_find_options_document()`, and `build_find_options_document()` doesn't add a read preference because it's not a find command option.

Would it be clearer if I renamed `build_find_options_document()` to `build_find_command_options_document()`?